certdb/ocspstapling.go

```diff
@@ -0,0 +1,153 @@
+	ocspRecs, err := acc.GetOCSP(serial, aki)
+	if err != nil || len(ocspRecs) == 0 {
+		// { There was an error or the OCSPRecord does not exist }
+		return false
+	}
+
+	// Add the SCTs to each ocsp response
+	for _, rec := range ocspRecs {
+		// NOTE: Assuming that rec.Body is in base64-encoded DER form
+		// based on
+		// https://github.com/cloudflare/cfssl/blob/master/ocsp/responder.go#L54
+		der, err := base64.StdEncoding.DecodeString(rec.Body)
+		if err != nil {
+			// { decoding error }
+			return false
+		}
+
+		response, err := ocsp.ParseResponse(der, nil)
+		if err != nil {
+			// { parsing error }
+			return false
+		}
+
+		// Serialize the list of SCTs
+		serializedSCTList, err := serializeSCTList(scts)
+		if err != nil {
+			// { serializing error }
+			return false
+		}
+
+		serializedSCTList, err = asn1.Marshal(serializedSCTList)
+		if err != nil {
+			// { serializing error }
+			return false
+		}
+
+		// Add the SCT list to a new extension
+		//
+		// The body of the extension should be a
+		// SignedCertificateTimestampList ::= OCTET STRING (see RFC
+		// 6962 sec 3.3)
+		sctExtension := pkix.Extension{
+			Id:       sctExtOid,
+			Critical: false,
+			Value:    serializedSCTList,
+		}
+
+		// Find the SCTListExtension in the ocsp response
+		var idxExt int
+		for _, ext := range response.Extensions {
+			if ext.Id.Equal(sctExtOid) {
+				break
+			}
+			idxExt++
+		}
+
+		if idxExt < len(response.Extensions) {
+			// { there's an existing SCT extension in response.Extensions }
+			// Replace the old extension in the OCSP response
+			response.Extensions[idxExt] = sctExtension
+		} else {
+			// { there's no SCT extension in response.Extensions }
+			response.Extensions = append(response.Extensions, sctExtension)
+		}
+
+		// Write updated extensions to replace existing extensions in
+		// response when re-marshalling
+		response.ExtraExtensions = response.Extensions
+
+		// Re-sign response to generate the new DER-encoded response
+		der, err = ocsp.CreateResponse(nil, response.Certificate, *response, priv)
+
+		if err != nil {
+			return false
+		}
+
+		body := base64.StdEncoding.EncodeToString(der)
+		err = acc.UpdateOCSP(serial, aki, body, rec.Expiry)
+
+		if err != nil {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Copied from
+// https://github.com/cloudflare/cfssl/blob/master/signer/local/local.go
+// because of dependency cycle
+func serializeSCTList(sctList []ct.SignedCertificateTimestamp) ([]byte, error) {
+	var buf bytes.Buffer
+	for _, sct := range sctList {
+		sct, err := ct.SerializeSCT(sct)
+		if err != nil {
+			return nil, err
+		}
+		binary.Write(&buf, binary.BigEndian, uint16(len(sct)))
+		buf.Write(sct)
+	}
+
+	var sctListLengthField = make([]byte, 2)
+	binary.BigEndian.PutUint16(sctListLengthField, uint16(buf.Len()))
+	return bytes.Join([][]byte{sctListLengthField, buf.Bytes()}, nil), nil
+}
+
+func deserializeSCTList(serializedSCTList []byte) ([]ct.SignedCertificateTimestamp, error) {
```

Comment:
so we don't need it here, but still, move it to helper function and make local.go use it as well.